File: ios/new/RiveReactNativeView.swift
import RiveRuntime
import NitroModules
import UIKit

enum BindData {
  case none
  case auto
  case instance(ViewModelInstance)
  case byName(String)
}

struct ViewConfiguration {
  let artboardName: String?
  let stateMachineName: String?
  let autoPlay: Bool
  let file: File
  let fit: RiveRuntime.Fit
  let semantics: RiveRuntime.Semantics
  let frameRate: RiveRuntime.FrameRate
  let offscreenBehavior: OffscreenBehavior
  let bindData: BindData
}

@MainActor
class RiveReactNativeView: UIView {
  private var riveUIView: RiveUIView?
  private var riveInstance: RiveRuntime.Rive?
  private var pendingBindInstance: ViewModelInstance?
  private var viewReadyContinuations: [CheckedContinuation<Bool, Never>] = []
  private var isViewReady = false
  private var configTask: Task<Void, Never>?
  private var settledTask: Task<Void, Never>?
  private var stopNotifyTask: Task<Void, Never>?
  private var isPaused = false
  private var semantics: RiveRuntime.Semantics = RiveUIView.Constants.Defaults.semantics {
    didSet { riveUIView?.semantics = semantics }
  }
  private var frameRate: RiveRuntime.FrameRate = RiveUIView.Constants.Defaults.frameRate {
    didSet { riveUIView?.frameRate = frameRate }
  }
  var autoPlay: Bool = true

  // The upstream runtime couples advancing and drawing behind a single
  // isPaused flag, so only .pause is implementable here — .skipDraws behaves
  // like .none (iOS already throttles most offscreen rendering on its own).
  // Never automatic — data-binding consumers may rely on the state machine
  // advancing while offscreen. Scrolling produces no callback on this view,
  // so visibility is polled on a low-frequency timer that only runs while
  // opted in.
  private var offscreenBehavior: OffscreenBehavior = .none {
    didSet {
      guard offscreenBehavior != oldValue else { return }
      updateVisibilityPolling()
    }
  }
  private var isOffscreen = false {
    didSet {
      guard isOffscreen != oldValue else { return }
      applyPauseState()
    }
  }
  private var visibilityTimer: Timer?

  private func updateVisibilityPolling() {
    visibilityTimer?.invalidate()
    visibilityTimer = nil
    guard offscreenBehavior == .pause else {
      isOffscreen = false
      return
    }
    guard window != nil else {
      isOffscreen = true
      return
    }
    isOffscreen = !isInVisibleViewport()
    let timer = Timer(timeInterval: 0.25, repeats: true) { [weak self] _ in
      MainActor.assumeIsolated {
        guard let self else { return }
        self.isOffscreen = !self.isInVisibleViewport()
      }
    }
    // .common so polling continues while the user is dragging a scroll view —
    // that's exactly when visibility changes.
    RunLoop.main.add(timer, forMode: .common)
    visibilityTimer = timer
  }

  private func isInVisibleViewport() -> Bool {
    guard let window, !isHidden, alpha > 0, bounds.width > 0, bounds.height > 0 else {
      return false
    }
    let frameInWindow = convert(bounds, to: window)
    return frameInWindow.intersects(window.bounds)
  }

  private func applyPauseState() {
    riveUIView?.isPaused = isPaused || (offscreenBehavior == .pause && isOffscreen)
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    updateVisibilityPolling()
  }

  /// Configure failures are reported here (wired to the onError prop).
  var onLoadError: ((String) -> Void)?

  /// Fired whenever the state machine settles (reaches rest, e.g. a
  /// non-looping animation reaching its end); wired to the onStop prop.
  var onSettled: (() -> Void)?

  func awaitViewReady() async -> Bool {
    if isViewReady {
      return true
    }
    return await withCheckedContinuation { continuation in
      viewReadyContinuations.append(continuation)
    }
  }

  private func resumeViewReadyContinuations(_ result: Bool) {
    let continuations = viewReadyContinuations
    viewReadyContinuations.removeAll()
    for continuation in continuations {
      continuation.resume(returning: result)
    }
  }

  func configure(_ config: ViewConfiguration, dataBindingChanged: Bool = false, reload: Bool = false, initialUpdate: Bool = false) {
    dispatchPrecondition(condition: .onQueue(.main))

    semantics = config.semantics
    frameRate = config.frameRate
    offscreenBehavior = config.offscreenBehavior

    if reload {
      cleanup()
    }

    // A dataBind-only change with a live instance re-binds the running state
    // machine instead of rebuilding artboard + state machine, which would
    // visually restart the animation. Only the .instance case can bind
    // without async resolution; the other modes fall through to the rebuild.
    if !reload, !initialUpdate, dataBindingChanged,
       let rive = riveInstance, case .instance(let vmi) = config.bindData {
      rive.stateMachine.bindViewModelInstance(vmi)
      rive.fit = config.fit
      return
    }

    if reload || dataBindingChanged || initialUpdate {
      // Applied synchronously so a play()/pause() issued while the config
      // task is still in flight isn't clobbered when it completes — the task
      // reads the then-current isPaused instead of recomputing it.
      isPaused = !config.autoPlay
      configTask?.cancel()
      configTask = Task { [weak self] in
        guard let self else { return }
        do {
          let artboard = try await config.file.createArtboard(config.artboardName)
          let stateMachine = try await artboard.createStateMachine(config.stateMachineName)

          let dataBind: RiveRuntime.DataBind
          switch config.bindData {
          case .none:
            dataBind = .none
          case .auto:
            // Probe for a default ViewModel first. If the artboard has none,
            // the SDK would fire an error event — skip auto-binding silently instead.
            do {
              let _ = try await config.file.getDefaultViewModelInfo(for: artboard)
              dataBind = .auto
            } catch {
              dataBind = .none
            }
          case .instance(let vmi):
            dataBind = .instance(vmi)
          case .byName(let name):
            let vmInfo = try await config.file.getDefaultViewModelInfo(for: artboard)
            let vmi = try await config.file.createViewModelInstance(.name(name, from: .name(vmInfo.viewModelName)))
            dataBind = .instance(vmi)
          }

          guard !Task.isCancelled else { return }

          let rive = try await RiveRuntime.Rive(
            file: config.file,
            artboard: artboard,
            stateMachine: stateMachine,
            dataBind: dataBind,
            fit: config.fit
          )

          guard !Task.isCancelled else { return }

          self.riveInstance = rive
          if let pending = self.pendingBindInstance {
            self.pendingBindInstance = nil
            rive.stateMachine.bindViewModelInstance(pending)
          }
          self.setupRiveUIView(with: rive)

          if !self.isViewReady {
            self.isViewReady = true
            self.resumeViewReadyContinuations(true)
          }
        } catch {
          guard !Task.isCancelled else { return }
          let message = (error as NSError).localizedDescription
          RiveLog.e("RiveReactNativeView", "Failed to configure: \(message)")
          self.onLoadError?(message)
          self.resumeViewReadyContinuations(false)
        }
      }
    } else {
      riveInstance?.fit = config.fit
    }
  }

  func bindViewModelInstance(viewModelInstance: ViewModelInstance) {
    if let rive = riveInstance {
      rive.stateMachine.bindViewModelInstance(viewModelInstance)
    } else {
      // configTask hasn't finished yet — queue it to apply once riveInstance is set
      pendingBindInstance = viewModelInstance
    }
  }

  func getViewModelInstance() -> ViewModelInstance? {
    return riveInstance?.viewModelInstance
  }

  func play() {
    isPaused = false
    applyPauseState()
  }

  func pause() {
    isPaused = true
    applyPauseState()
  }

  func reset() {
    // Deprecated: the experimental Rive runtime has no reset primitive.
    RiveLog.e("RiveReactNativeView", "reset() is deprecated and not supported on the new runtime")
  }

  func playIfNeeded() {
    if isPaused {
      isPaused = false
      applyPauseState()
    }
  }

  func setNumberInputValue(name: String, value: Float, path: String?) throws {
    throw RuntimeError.error(withMessage: "SMI inputs are not supported by the new runtime — use data binding")
  }

  func getNumberInputValue(name: String, path: String?) throws -> Float {
    throw RuntimeError.error(withMessage: "SMI inputs are not supported by the new runtime — use data binding")
  }

  func setBooleanInputValue(name: String, value: Bool, path: String?) throws {
    throw RuntimeError.error(withMessage: "SMI inputs are not supported by the new runtime — use data binding")
  }

  func getBooleanInputValue(name: String, path: String?) throws -> Bool {
    throw RuntimeError.error(withMessage: "SMI inputs are not supported by the new runtime — use data binding")
  }

  func triggerInput(name: String, path: String?) throws {
    throw RuntimeError.error(withMessage: "SMI inputs are not supported by the new runtime — use data binding")
  }

  func setTextRunValue(name: String, value: String, path: String?) throws {
    throw RuntimeError.error(withMessage: "Text runs are not supported by the new runtime — use data binding")
  }

  func getTextRunValue(name: String, path: String?) throws -> String {
    throw RuntimeError.error(withMessage: "Text runs are not supported by the new runtime — use data binding")
  }

  // MARK: - Internal

  private func observeSettled(of rive: RiveRuntime.Rive) {
    settledTask?.cancel()
    // A stop scheduled by the previous Rive instance must not fire into the
    // new one after a reconfigure.
    stopNotifyTask?.cancel()
    stopNotifyTask = nil
    settledTask = Task { [weak self] in
      for await _ in rive.stateMachine.settledStream() {
        guard !Task.isCancelled else { return }
        self?.scheduleStopNotification()
      }
    }
  }

  // The SDK can emit settle back-to-back right after configure (an initial
  // rest, immediately perturbed by setup — e.g. data-bind — and re-settling),
  // which would otherwise report two stops for a single, user-visible "came
  // to rest". Coalesce a tight burst into one onStop call.
  private func scheduleStopNotification() {
    stopNotifyTask?.cancel()
    stopNotifyTask = Task { [weak self] in
      try? await Task.sleep(nanoseconds: 150_000_000)
      guard !Task.isCancelled else { return }
      self?.onSettled?()
    }
  }

  private func setupRiveUIView(with rive: RiveRuntime.Rive) {
    observeSettled(of: rive)
    if let existing = riveUIView {
      // Reuse the existing view — avoids tearing down the MTKView on every
      // reconfigure, which previously caused orphaned draw calls ("state machine
      // not found") from the old MTKView after removeFromSuperview.
      existing.rive = rive
      existing.semantics = semantics
      existing.frameRate = frameRate
      applyPauseState()
    } else {
      let uiView = RiveUIView(rive: rive, isPaused: isPaused)
      uiView.semantics = semantics
      uiView.frameRate = frameRate
      uiView.translatesAutoresizingMaskIntoConstraints = false
      addSubview(uiView)
      NSLayoutConstraint.activate([
        uiView.leadingAnchor.constraint(equalTo: leadingAnchor),
        uiView.trailingAnchor.constraint(equalTo: trailingAnchor),
        uiView.topAnchor.constraint(equalTo: topAnchor),
        uiView.bottomAnchor.constraint(equalTo: bottomAnchor),
      ])
      self.riveUIView = uiView
      applyPauseState()
    }
  }

  private func cleanup() {
    dispatchPrecondition(condition: .onQueue(.main))
    configTask?.cancel()
    configTask = nil
    settledTask?.cancel()
    settledTask = nil
    stopNotifyTask?.cancel()
    stopNotifyTask = nil
    riveUIView?.removeFromSuperview()
    riveUIView = nil
    riveInstance = nil
    pendingBindInstance = nil
  }

  /// Final teardown, called from HybridRiveView.dispose() (always on main).
  /// Unlike the reload-path cleanup(), this also settles any pending
  /// awaitViewReady() callers so their promises (which retain this view)
  /// don't hang forever.
  func detach() {
    visibilityTimer?.invalidate()
    visibilityTimer = nil
    resumeViewReadyContinuations(false)
    cleanup()
  }

  deinit {
    // deinit is nonisolated and Nitro can drop the last reference off the
    // main thread (JS/GC thread); cleanup() must run on main. Capture the
    // resources, not self.
    let task = configTask
    let settled = settledTask
    let stopNotify = stopNotifyTask
    let uiView = riveUIView
    let timer = visibilityTimer
    if Thread.isMainThread {
      timer?.invalidate()
      task?.cancel()
      settled?.cancel()
      stopNotify?.cancel()
      uiView?.removeFromSuperview()
    } else {
      DispatchQueue.main.async {
        timer?.invalidate()
        task?.cancel()
        settled?.cancel()
        stopNotify?.cancel()
        uiView?.removeFromSuperview()
      }
    }
  }
}

File: android/src/new/java/com/rive/RiveReactNativeView.kt
package com.rive

import android.annotation.SuppressLint
import android.graphics.Rect
import android.graphics.SurfaceTexture
import android.os.Build
import android.util.Log
import android.view.Choreographer
import android.view.MotionEvent
import android.view.TextureView
import android.view.View
import android.widget.FrameLayout
import com.facebook.react.bridge.UiThreadUtil
import app.rive.Artboard
import app.rive.Fit
import app.rive.RiveFile
import app.rive.ViewModelInstance
import app.rive.ViewModelSource
import app.rive.core.ArtboardHandle
import app.rive.core.CommandQueue
import app.rive.core.RiveSurface
import app.rive.core.StateMachineHandle
import app.rive.core.SurfaceTextureSurface
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.rive.OffscreenBehavior
import com.margelo.nitro.rive.RiveErrorLogger
import com.margelo.nitro.rive.RiveLog
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.time.Duration
import kotlin.time.Duration.Companion.nanoseconds

// The renderEnabled prop (boolean | 'pause') resolved into what the render
// loop should do: draw normally, keep advancing but skip draws, or stop both.
enum class RenderMode { Enabled, SkipDraws, Paused }

sealed class BindData {
  data object None : BindData()
  data object Auto : BindData()
  data class Instance(val instance: ViewModelInstance) : BindData()
  data class ByName(val name: String) : BindData()
}

data class ViewConfiguration(
  val artboardName: String?,
  val stateMachineName: String?,
  val autoPlay: Boolean,
  val riveFile: RiveFile,
  val riveWorker: CommandQueue,
  val alignment: app.rive.Alignment,
  val fit: app.rive.Fit,
  val layoutScaleFactor: Float?,
  val bindData: BindData
)

@SuppressLint("ViewConstructor")
class RiveReactNativeView(context: ThemedReactContext) : FrameLayout(context) {
  companion object {
    private const val TAG = "RiveReactNativeView"

    // Half of a 120Hz vsync: lets a cap land on the nearest vsync multiple
    // (e.g. every 4th frame at 120Hz for a 30fps cap) instead of drifting past
    // it and halving the effective rate.
    private const val CAP_TOLERANCE_NS = 4_000_000L
  }

  // Render at most this many frames per second; null = every vsync.
  var frameRate: Double? = null
    set(value) {
      field = value
      updateFrameRateHint()
    }

  // What to do while the view is outside the visible viewport (scrolled out,
  // hidden, or windowless): keep running, skip draws only, or pause fully.
  // Never automatic — data-binding consumers may rely on the state machine
  // advancing while offscreen. Overlays in the same window (e.g. RN Modal)
  // don't count as covering the view.
  var offscreenBehavior: OffscreenBehavior = OffscreenBehavior.NONE

  // Manual counterpart to offscreenBehavior for occlusion the view can't
  // detect (RN Modal, bottom sheets): SkipDraws keeps the state machine
  // advancing, Paused stops it too (composes with the pause()/play() state
  // rather than overwriting it).
  var renderMode: RenderMode = RenderMode.Enabled

  private val visibleRectBuffer = Rect()

  private fun isInVisibleViewport(): Boolean =
    isShown && getGlobalVisibleRect(visibleRectBuffer)

  var onError: ((String) -> Unit)? = null

  // Fired when the state machine settles (reaches rest, e.g. a non-looping
  // animation reaching its end); wired to the onStop prop.
  var onStop: (() -> Unit)? = null

  private var settledJob: Job? = null

  // rive-runtime's command server emits a settle signal on every advance
  // whose advanceAndApply returns false, not just on the settle edge — so
  // once the state machine is at rest we simply stop advancing it (also
  // saves CPU). Re-armed by whatever could actually move the state machine
  // again: pointer input, resuming playback, or a data-binding change.
  @Volatile
  private var settled = false

  private val errorListener: (String) -> Unit = { msg ->
    onError?.invoke(msg)
  }

  private val viewReadyDeferred = CompletableDeferred<Boolean>()
  private var boundInstance: ViewModelInstance? = null

  // Instances created by the view itself (Auto/ByName binding) must be closed
  // by the view; instances passed in from JS are owned by their JS wrapper.
  private var ownsBoundInstance = false
  private var riveWorker: CommandQueue? = null
  private var activeFit: Fit = Fit.Contain()

  private var riveFile: RiveFile? = null
  private var artboard: Artboard? = null
  private var artboardHandle: ArtboardHandle? = null
  private var stateMachineHandle: StateMachineHandle? = null
  private var riveSurface: RiveSurface? = null

  private var surfaceTexture: SurfaceTexture? = null
  private var surfaceWidth = 0
  private var surfaceHeight = 0

  private var renderLoopRunning = false
  private var disposed = false
  private var lastFrameTimeNs = 0L
  private var frameCount = 0L

  // While paused the loop still ticks, but only draws when something changed
  // (initial content, resize, rebinding) — otherwise a paused view would keep
  // re-rendering identical frames at full refresh rate.
  private var needsRedraw = true

  @Volatile
  private var paused = false

  private val viewScope = CoroutineScope(Dispatchers.Default + SupervisorJob())

  private val textureView = TextureView(context).apply {
    layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    surfaceTextureListener = object : TextureView.SurfaceTextureListener {
      override fun onSurfaceTextureAvailable(st: SurfaceTexture, w: Int, h: Int) {
        Log.d(TAG, "onSurfaceTextureAvailable: ${w}x$h worker=${this@RiveReactNativeView.riveWorker != null}")
        this@RiveReactNativeView.surfaceTexture = st
        this@RiveReactNativeView.surfaceWidth = w
        this@RiveReactNativeView.surfaceHeight = h
        this@RiveReactNativeView.riveWorker?.let { worker ->
          if (this@RiveReactNativeView.riveSurface == null) {
            this@RiveReactNativeView.riveSurface = worker.createRiveSurface(SurfaceTextureSurface(st, w, h))
            Log.d(TAG, "onSurfaceTextureAvailable: surface created")
            resizeArtboardIfLayout()
          }
        }
      }

      override fun onSurfaceTextureDestroyed(st: SurfaceTexture): Boolean {
        this@RiveReactNativeView.riveSurface?.let { surface ->
          runCatching { surface.close() }
        }
        this@RiveReactNativeView.riveSurface = null
        return false
      }

      override fun onSurfaceTextureSizeChanged(st: SurfaceTexture, w: Int, h: Int) {
        this@RiveReactNativeView.surfaceWidth = w
        this@RiveReactNativeView.surfaceHeight = h
        this@RiveReactNativeView.needsRedraw = true
        // Since 11.7.x the render target keeps its creation-time size and
        // RiveSurface.resize() is internal to the SDK, so only the artboard
        // is resized here (same behavior as before the 11.7.2 bump).
        resizeArtboardIfLayout()
      }

      override fun onSurfaceTextureUpdated(st: SurfaceTexture) {}
    }
  }

  init {
    addView(textureView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
  }

  private val renderCallback = object : Choreographer.FrameCallback {
    override fun doFrame(frameTimeNanos: Long) {
      if (!renderLoopRunning || disposed) return

      val offscreen = offscreenBehavior != OffscreenBehavior.NONE && !isInVisibleViewport()
      val pausedOffscreen = offscreen && offscreenBehavior == OffscreenBehavior.PAUSE
      val renderPaused = renderMode == RenderMode.Paused

      if ((paused || pausedOffscreen || renderPaused) && !needsRedraw) {
        // Keep the timebase fresh so resuming advances by one frame, not by
        // the whole pause span.
        lastFrameTimeNs = frameTimeNanos
        Choreographer.getInstance().postFrameCallback(this)
        return
      }

      val capPeriodNs = frameRate
        ?.takeIf { it > 0 }
        ?.let { (1_000_000_000.0 / it).toLong() }
      if (capPeriodNs != null && lastFrameTimeNs != 0L &&
        frameTimeNanos - lastFrameTimeNs < capPeriodNs - CAP_TOLERANCE_NS
      ) {
        // Skip without touching lastFrameTimeNs: the eventual advance must
        // cover the full elapsed time so capped playback keeps wall-clock speed.
        Choreographer.getInstance().postFrameCallback(this)
        return
      }

      val deltaTime = if (lastFrameTimeNs == 0L) {
        Duration.ZERO
      } else {
        (frameTimeNanos - lastFrameTimeNs).nanoseconds
      }
      lastFrameTimeNs = frameTimeNanos

      val worker = riveWorker
      val art = artboardHandle
      val sm = stateMachineHandle
      val rs = riveSurface

      // Offscreen views keep advancing the state machine (events and
      // data-binding listeners stay live) and only skip the draw — the draw
      // is the dominant part of the offscreen cost. needsRedraw still forces
      // a draw so pending content (initial frame, resize, rebinding) isn't
      // lost while invisible.
      val skipDraw = renderMode != RenderMode.Enabled || offscreen

      if (worker != null && art != null && sm != null && rs != null) {
        try {
          if (!paused && !settled && !pausedOffscreen && !renderPaused) {
            worker.advanceStateMachine(sm, deltaTime)
          }
          if (!skipDraw || needsRedraw) {
            worker.draw(art, sm, rs, activeFit)
            needsRedraw = false
            frameCount++
            val isFirstFrame = frameCount == 1L
            if (isFirstFrame) {
              viewReadyDeferred.complete(true)
            }
          }
        } catch (e: Exception) {
          Log.e(TAG, "Render loop error", e)
        }
      }

      if (!disposed) {
        Choreographer.getInstance().postFrameCallback(this)
      }
    }
  }

  private fun startRenderLoop() {
    if (renderLoopRunning) return
    renderLoopRunning = true
    lastFrameTimeNs = 0L
    updateFrameRateHint()
    Choreographer.getInstance().postFrameCallback(renderCallback)
  }

  private fun stopRenderLoop() {
    renderLoopRunning = false
    updateFrameRateHint()
    Choreographer.getInstance().removeFrameCallback(renderCallback)
  }

  // Advisory platform hint (upstream applies the same one inside its Compose
  // loop): on capable displays a capped, actively-drawing view lets the
  // system lower the refresh rate, saving power beyond the skipped draws.
  // Callers may be off-main (play()/pause() run on a coroutine), so hop.
  private fun updateFrameRateHint() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.VANILLA_ICE_CREAM) return
    UiThreadUtil.runOnUiThread {
      val fps = frameRate?.takeIf { it > 0 && renderLoopRunning && !paused }
      textureView.requestedFrameRate =
        fps?.toFloat() ?: View.REQUESTED_FRAME_RATE_CATEGORY_NO_PREFERENCE
    }
  }

  suspend fun awaitViewReady(): Boolean {
    return viewReadyDeferred.await()
  }

  fun configure(config: ViewConfiguration, dataBindingChanged: Boolean, reload: Boolean = false, initialUpdate: Boolean = false) {
    riveWorker = config.riveWorker
    activeFit = config.fit
    needsRedraw = true
    Log.d(
      TAG,
      "configure: reload=$reload initialUpdate=$initialUpdate fit=$activeFit surfaceTexture=${surfaceTexture != null} surfaceW=$surfaceWidth surfaceH=$surfaceHeight"
    )

    if (reload) {
      RiveErrorLogger.resetReportedErrors()
      RiveErrorLogger.addListener(errorListener)
      stateMachineHandle?.let { old ->
        runCatching { config.riveWorker.deleteStateMachine(old) }
      }
      stateMachineHandle = null
      artboard?.close()

      val newArtboard = if (config.artboardName != null) {
        Artboard.fromFile(config.riveFile, config.artboardName)
      } else {
        Artboard.fromFile(config.riveFile)
      }
      artboard = newArtboard
      artboardHandle = newArtboard.artboardHandle

      riveFile = config.riveFile

      val newStateMachineHandle = if (config.stateMachineName != null) {
        config.riveWorker.createStateMachineByName(newArtboard.artboardHandle, config.stateMachineName)
      } else {
        config.riveWorker.createDefaultStateMachine(newArtboard.artboardHandle)
      }
      stateMachineHandle = newStateMachineHandle
      observeSettled(config.riveWorker, newStateMachineHandle)

      if (surfaceTexture != null && riveSurface == null) {
        riveSurface = config.riveWorker.createRiveSurface(
          SurfaceTextureSurface(surfaceTexture!!, surfaceWidth, surfaceHeight)
        )
      }

      Log.d(TAG, "configure: artboard=${artboardHandle != null} sm=${stateMachineHandle != null} surface=${riveSurface != null}")

      paused = !config.autoPlay
      startRenderLoop()
    }

    resizeArtboardIfLayout()

    if (dataBindingChanged || initialUpdate || reload) {
      applyDataBinding(config.bindData, config.riveFile)
    }
  }

  private fun observeSettled(worker: CommandQueue, handle: StateMachineHandle) {
    settled = false
    settledJob?.cancel()
    settledJob = viewScope.launch {
      worker.settledFlow.collect { settledHandle ->
        if (settledHandle == handle && !settled) {
          settled = true
          onStop?.invoke()
        }
      }
    }
  }

  private fun resizeArtboardIfLayout() {
    val fit = activeFit
    if (fit is Fit.Layout) {
      val rs = riveSurface ?: return
      val art = artboard ?: return
      art.resizeArtboard(rs, fit.scaleFactor)
    }
  }

  override fun onInterceptTouchEvent(ev: MotionEvent?): Boolean = true

  @SuppressLint("ClickableViewAccessibility")
  override fun onTouchEvent(event: MotionEvent): Boolean {
    handlePointerEvent(event)
    return true
  }

  private fun handlePointerEvent(event: MotionEvent) {
    val worker = riveWorker ?: run {
      Log.w(TAG, "touch: no worker")
      return
    }
    val smHandle = stateMachineHandle ?: run {
      Log.w(TAG, "touch: no smHandle")
      return
    }
    val w = surfaceWidth.toFloat()
    val h = surfaceHeight.toFloat()
    if (w <= 0 || h <= 0) {
      Log.w(TAG, "touch: invalid surface ${w}x$h")
      return
    }

    val fit = activeFit

    try {
      when (event.actionMasked) {
        MotionEvent.ACTION_DOWN -> {
          worker.pointerDown(smHandle, fit, w, h, event.getPointerId(event.actionIndex), event.x, event.y)
        }
        MotionEvent.ACTION_MOVE -> {
          worker.pointerMove(smHandle, fit, w, h, event.getPointerId(0), event.x, event.y)
        }
        MotionEvent.ACTION_UP -> {
          val id = event.getPointerId(event.actionIndex)
          worker.pointerUp(smHandle, fit, w, h, id, event.x, event.y)
          worker.pointerExit(smHandle, fit, w, h, id, event.x, event.y)
        }
        MotionEvent.ACTION_CANCEL -> {
          val id = event.getPointerId(event.actionIndex)
          worker.pointerUp(smHandle, fit, w, h, id, -1f, -1f)
          worker.pointerExit(smHandle, fit, w, h, id, -1f, -1f)
        }
      }
      // A pointer event may move the state machine off rest; resume advancing
      // so the render loop actually applies its effect.
      settled = false
    } catch (e: Exception) {
      Log.e(TAG, "Pointer event failed", e)
    }
  }

  fun bindViewModelInstance(vmi: ViewModelInstance) {
    boundInstance = vmi
  }

  fun getViewModelInstance(): ViewModelInstance? {
    return boundInstance
  }

  private fun setBoundInstance(instance: ViewModelInstance?, owns: Boolean) {
    val previous = boundInstance
    if (ownsBoundInstance && previous != null && previous !== instance) {
      runCatching { previous.close() }
    }
    boundInstance = instance
    ownsBoundInstance = owns
  }

  private fun applyDataBinding(bindData: BindData, riveFile: RiveFile) {
    when (bindData) {
      is BindData.None -> {
        setBoundInstance(null, owns = false)
      }
      is BindData.Auto -> {
        viewScope.launch {
          try {
            val vmNames = riveFile.getViewModelNames()
            if (vmNames.isEmpty()) return@launch
            withContext(Dispatchers.Main) {
              if (disposed) return@withContext
              val art = artboard ?: return@withContext
              // Probe for a default ViewModel first — getDefaultViewModelInfo
              // throws when the artboard has none, a normal state. (Creating
              // the instance regardless and checking its handle against a
              // magic value relied on undocumented handle allocation.)
              try {
                riveFile.getDefaultViewModelInfo(art)
              } catch (e: Exception) {
                Log.d(TAG, "Auto-binding skipped: no default ViewModel for artboard")
                return@withContext
              }
              if (disposed) return@withContext
              val source = ViewModelSource.DefaultForArtboard(art).defaultInstance()
              val instance = ViewModelInstance.fromFile(riveFile, source)
              setBoundInstance(instance, owns = true)
              bindInstanceToStateMachine(instance)
            }
          } catch (e: Exception) {
            Log.d(TAG, "Auto-binding skipped: ${e.message}")
          }
        }
      }
      is BindData.Instance -> {
        setBoundInstance(bindData.instance, owns = false)
        bindInstanceToStateMachine(bindData.instance)
      }
      is BindData.ByName -> {
        try {
          val art = artboard ?: return
          val source = ViewModelSource.DefaultForArtboard(art).namedInstance(bindData.name)
          val instance = ViewModelInstance.fromFile(riveFile, source)
          setBoundInstance(instance, owns = true)
          bindInstanceToStateMachine(instance)
        } catch (e: Exception) {
          Log.e(TAG, "Failed to create named instance", e)
        }
      }
    }
  }

  private fun bindInstanceToStateMachine(instance: ViewModelInstance) {
    val worker = riveWorker
    val smHandle = stateMachineHandle
    if (worker != null && smHandle != null) {
      worker.bindViewModelInstance(smHandle, instance.instanceHandle)
      needsRedraw = true
      settled = false
    } else {
      Log.w(TAG, "Cannot bind VMI: worker or state machine handle not available")
    }
  }

  fun play() {
    paused = false
    settled = false
    updateFrameRateHint()
  }

  fun pause() {
    paused = true
    updateFrameRateHint()
  }

  // Deprecated: the experimental Rive runtime has no reset primitive.
  fun reset() {
    RiveLog.e(TAG, "reset() is deprecated and not supported on the experimental backend")
  }

  fun playIfNeeded() {
    paused = false
    settled = false
    updateFrameRateHint()
  }

  fun setNumberInputValue(name: String, value: Double, path: String?) {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  fun getNumberInputValue(name: String, path: String?): Double {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  fun setBooleanInputValue(name: String, value: Boolean, path: String?) {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  fun getBooleanInputValue(name: String, path: String?): Boolean {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  fun triggerInput(name: String, path: String?) {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  fun setTextRunValue(name: String, value: String, path: String?) {
    throw UnsupportedOperationException("Text runs not supported in experimental API")
  }

  fun getTextRunValue(name: String, path: String?): String {
    throw UnsupportedOperationException("Text runs not supported in experimental API")
  }

  // Runs twice per view: eagerly from HybridRiveView.dispose() (JS unmount,
  // posted to main) and from RiveViewManager.onDropViewInstance — must stay
  // idempotent and on the main thread (Choreographer is thread-local).
  fun dispose() {
    if (disposed) return
    disposed = true
    viewReadyDeferred.complete(false)
    settledJob?.cancel()
    viewScope.cancel()
    RiveErrorLogger.removeListener(errorListener)
    stopRenderLoop()
    // The command queue is FIFO, so deletes enqueued here run after any
    // still-pending draw commands that reference these handles.
    setBoundInstance(null, owns = false)
    stateMachineHandle?.let { old ->
      riveWorker?.let { worker -> runCatching { worker.deleteStateMachine(old) } }
    }
    artboard?.let { runCatching { it.close() } }
    riveSurface?.let { runCatching { it.close() } }
    artboard = null
    artboardHandle = null
    stateMachineHandle = null
    riveSurface = null
  }
}
